### PR TITLE
Download components by component type

### PR DESCRIPTION
**Pull request by mlohvynenko**

### CODE REVIEW

File: utils/semverutils/semverutils.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package launcher provides set of API to controls services lifecycle
+
+package semverutils
```
Comment (al1img) [inline utils/semverutils/semverutils.go:20]: Should be part of aos_common/utils

Comment (mlohvynenko) [reply utils/semverutils/semverutils.go:20]: no objections at all
Anyway, I borrowed this util from Mykola Kobets PR https://github.com/aosedge/aos_communicationmanager/pull/129/files

So once that PR is merged, I will simply remove this WIP commit

File: unitstatushandler/firmwaremanager.go [hunk 2]
```diff
@@ -652,3 +746,82 @@ func (manager *firmwareManager) asyncUpdate(updateComponents []cloudprotocol.Com
 
 	return finishChannel
 }
+
+func getDownloadID(component cloudprotocol.ComponentInfo) string {
+	if component.ComponentType != "" {
```
Comment (al1img) [inline unitstatushandler/firmwaremanager.go:751]: We should not have the case where `component.ComponentType == ""`

File: unitstatushandler/firmwaremanager.go [hunk 3]
```diff
@@ -652,3 +746,82 @@ func (manager *firmwareManager) asyncUpdate(updateComponents []cloudprotocol.Com
...
+
+func handleDesiredComponentsWithNoID(componentsWithNoID []cloudprotocol.ComponentInfo,
+	installedComponents []cloudprotocol.ComponentStatus, update *firmwareUpdate,
+) {
+	log.Debug("Handle desired components with no ID")
+
+	newComponents := make(map[string]cloudprotocol.ComponentInfo)
+
+	for _, installedComponent := range installedComponents {
+		if containsComponent(update.Components, installedComponent.ComponentID) {
```
Comment (al1img) [inline unitstatushandler/firmwaremanager.go:782]: Use `slices.Contains` method

File: umcontroller/umcontroller.go [hunk 4]
```diff
@@ -72,12 +72,13 @@ type Controller struct {
 
 // ComponentStatus information about system component update.
 type ComponentStatus struct {
-	ID          string `json:"id"`
-	Version     string `json:"version"`
-	Annotations string `json:"annotations,omitempty"`
-	URL         string `json:"url"`
-	Sha256      []byte `json:"sha256"`
-	Size        uint64 `json:"size"`
+	ID            string `json:"id"`
```
Comment (al1img) [inline umcontroller/umcontroller.go:75]: `ComponentID`?

**Review by MykolaSolyanko — APPROVED** [review]

File: unitstatushandler/firmwaremanager.go [hunk 5]
```diff
@@ -575,6 +640,40 @@ func (manager *firmwareManager) sendCurrentStatus() {
...
 
+func (manager *firmwareManager) updateComponentStatusByDownloadID(id, status string,
+	componentErr *cloudprotocol.ErrorInfo,
+) {
+	manager.statusMutex.Lock()
+	defer manager.statusMutex.Unlock()
+
+	found := false
+
+	for _, info := range manager.ComponentStatuses {
```
[the remark this answers is not in the document]
Comment (mlohvynenko) [reply unitstatushandler/firmwaremanager.go:651]: IndexFunc returns the first index i satisfying f(s[i]), or -1 if none do.

In this change multiple components with different ComponentID might have  the  same dowloadID (componentType + version).
Logic updates status for all components that match downloadID.
So I don't see how  IndexFunc can simplify the code.

[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply unitstatushandler/firmwaremanager.go:651]: ok, clear

**Review by al1img — APPROVED** [review]

**Pull request merged**